In [9]:
import pandas

In [10]:
pandas.__version__

'1.4.4'

In [11]:
import pandas as pd

In [12]:
pd.__version__

'1.4.4'

In [13]:
url = ("https://raw.githubusercontent.com/andrewD46""/pandas_numpy_tasks/main/data/adult.data.csv")

In [14]:
adults = pd.read_csv(url)

In [15]:
adults

age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-managerial        Husband  White    Male   
2                Divorced  Handlers-cleaners  Not-in-family  White    Male   
3      Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4      Married-civ-spouse     Prof-specialty           Wife  Black  Female   
...                   ...                ...            ...    ...     ...   
32556  Married-civ-spouse       Tech-support           Wife  White  Female   
32557  Married-civ-spouse  Machine-op-inspct        Husband  White    Male   
32558             Widowed       Adm-clerical      Unmarried  White  Female   
32559       Never-married       Adm-clerical      Own-child  White    Male   
32560  Married-civ-spouse    Exec-managerial           Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country salary  
0              2174             0              40  United-States  <=50K  
1                 0             0              13  United-States  <=50K  
2                 0             0              40  United-States  <=50K  
3                 0             0              40  United-States  <=50K  
4                 0             0              40           Cuba  <=50K  
...             ...           ...             ...            ...    ...  
32556             0             0              38  United-States  <=50K  
32557             0             0              40  United-States   >50K  
32558             0             0              40  United-States  <=50K  
32559             0             0              20  United-States  <=50K  
32560         15024             0              40  United-States   >50K  

[32561 rows x 15 columns]

# 1. Посчитайте, сколько мужчин и женщин (признак sex) представлено в этом датасете

In [16]:
print("В датасете представлено {} мужчин и {} женщин.".format(sum(adults['sex'] == 'Male'), 
                                            sum(adults['sex'] == 'Female')))

В датасете представлено 21790 мужчин и 10771 женщин.


In [17]:
adults.groupby('sex').size()

sex
Female    10771
Male      21790
dtype: int64

# 2. Каков средний возраст мужчин (признак age) по всему датасету?

In [18]:
import numpy as np

In [19]:
adults.groupby("sex").agg({"age": np.mean, "sex": np.size})

age    sex
sex                     
Female  36.858230  10771
Male    39.433547  21790

In [20]:
adults.loc[adults["sex"] =="Male", "age"].mean()

39.43354749885268

# 3. Какова доля граждан Соединенных Штатов (признак native-country)?

In [21]:
float((adults['native-country'] == 'United-States').sum()) / adults.shape[0]

0.895857006848684

# 4-5. Рассчитайте среднее значение и среднеквадратичное отклонение возраста тех, кто получает более 50K в год (признак salary) и тех, кто получает менее 50K в год

In [22]:
age_1 = adults.loc[adults['salary'] == '>50K', 'age']
age_2 = adults.loc[adults['salary'] == '<=50K', 'age']


In [23]:
round(age_1.mean()), round(age_1.std(), 1)

(44, 10.5)

In [24]:
round(age_2.mean()), round(age_2.std(), 1)

(37, 14.0)

# 6. Правда ли, что люди, которые получают больше 50k, имеют минимум высшее образование? (признак education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters или Doctorate)

In [25]:
adults.loc[adults['salary'] == '>50K', 'education'].unique() 

array(['HS-grad', 'Masters', 'Bachelors', 'Some-college', 'Assoc-voc',
       'Doctorate', 'Prof-school', 'Assoc-acdm', '7th-8th', '12th',
       '10th', '11th', '9th', '5th-6th', '1st-4th'], dtype=object)

Правда ли, что люди, которые получают больше 50k, имеют минимум высшее образование? NO

# 7. Выведите статистику возраста для каждой расы (признак race) и каждого пола. Используйте groupby и describe. Найдите таким образом максимальный возраст мужчин расы Asian-Pac-Islander.

In [26]:
adults.groupby(['race','sex'])['age'].describe()

count       mean        std   min   25%   50%  \
race               sex                                                       
Amer-Indian-Eskimo Female    119.0  37.117647  13.114991  17.0  27.0  36.0   
                   Male      192.0  37.208333  12.049563  17.0  28.0  35.0   
Asian-Pac-Islander Female    346.0  35.089595  12.300845  17.0  25.0  33.0   
                   Male      693.0  39.073593  12.883944  18.0  29.0  37.0   
Black              Female   1555.0  37.854019  12.637197  17.0  28.0  37.0   
                   Male     1569.0  37.682600  12.882612  17.0  27.0  36.0   
Other              Female    109.0  31.678899  11.631599  17.0  23.0  29.0   
                   Male      162.0  34.654321  11.355531  17.0  26.0  32.0   
White              Female   8642.0  36.811618  14.329093  17.0  25.0  35.0   
                   Male    19174.0  39.652498  13.436029  17.0  29.0  38.0   

                             75%   max  
race               sex                  
Amer-Indian-Eskimo Female  46.00  80.0  
                   Male    45.00  82.0  
Asian-Pac-Islander Female  43.75  75.0  
                   Male    46.00  90.0  
Black              Female  46.00  90.0  
                   Male    46.00  90.0  
Other              Female  39.00  74.0  
                   Male    42.00  77.0  
White              Female  46.00  90.0  
                   Male    49.00  90.0

In [27]:
adults[(adults.race == 'Asian-Pac-Islander')& (adults.sex == 'Male')].max()

age                               90
workclass                Without-pay
fnlwgt                        506329
education               Some-college
education-num                     16
marital-status               Widowed
occupation          Transport-moving
relationship               Unmarried
race              Asian-Pac-Islander
sex                             Male
capital-gain                   99999
capital-loss                    2457
hours-per-week                    99
native-country               Vietnam
salary                          >50K
dtype: object

# 8. Среди кого больше доля зарабатывающих много (>50K): среди женатых или холостых мужчин (признак marital-status)? Женатыми считаем тех, у кого marital-status начинается с Married (Married-civ-spouse, Married-spouse-absent или Married-AF-spouse), остальных считаем холостыми.

In [28]:
adults.loc[(adults['sex'] == 'Male') &
     (adults['marital-status'].isin(['Never-married', 'Separated','Divorced','Widowed'])), 'salary'].value_counts()

<=50K    7552
>50K      697
Name: salary, dtype: int64

In [29]:
adults.loc[(adults['sex'] == 'Male') & (adults['marital-status'].str.startswith('Married')), 'salary'].value_counts()


<=50K    7576
>50K     5965
Name: salary, dtype: int64

Доля зарабатывающих много (>50K) больше среди женатых 5965 чем среди холостых 697 мужчин

# 9. Какое максимальное число часов человек работает в неделю (признак hours-per-week)? Сколько людей работают такое количество часов и каков среди них процент зарабатывающих много?

In [30]:
adults['hours-per-week'].describe()['max']

99.0

In [31]:
a = adults[adults['hours-per-week'] == 99]
print(len(a))

85


# 10. Посчитайте среднее время работы (hours-per-week) зарабатывающих мало и много (salary) для каждой страны (native-country).

In [32]:
adults.groupby(['native-country', 'salary'])['hours-per-week'].mean()

native-country  salary
?               <=50K     40.164760
                >50K      45.547945
Cambodia        <=50K     41.416667
                >50K      40.000000
Canada          <=50K     37.914634
                            ...    
United-States   >50K      45.505369
Vietnam         <=50K     37.193548
                >50K      39.200000
Yugoslavia      <=50K     41.600000
                >50K      49.500000
Name: hours-per-week, Length: 82, dtype: float64

# 11.Сгруппируйте людей по возрастным группам young, adult, retiree, где:

young соответствует 16-35 лет
adult - 35-70 лет
retiree - 70-100 лет
Проставьте название соответсвтуещей группы для каждого человека в новой колонке AgeGroup

In [33]:
adults.groupby(['age']).mean()

fnlwgt  education-num  capital-gain  capital-loss  hours-per-week
age                                                                          
17   185722.232911       6.648101    109.691139     41.751899       21.367089
18   193530.207273       8.152727    102.058182     44.336364       25.912727
19   204803.543539       9.096910     92.092697     43.915730       30.678371
20   198537.705179       9.345286     82.786189     33.298805       32.280212
21   200116.997222       9.434722    204.625000     51.636111       34.034722
..             ...            ...           ...           ...             ...
85   126667.333333      10.333333      0.000000      0.000000       29.333333
86   149912.000000      14.000000      0.000000      0.000000       40.000000
87    90338.000000       9.000000      0.000000      0.000000        2.000000
88   153975.666667      13.333333      0.000000      0.000000       40.000000
90   161458.604651       9.883721   1935.255814    152.604651       36.813953

[73 rows x 5 columns]

In [34]:
adults['age'] = adults['age'].astype(str).str.strip('+').astype(int)

ad  = adults.join(pd.get_dummies(pd.cut(adults['age'],
                           bins=(16,35,70,100,np.inf), 
                           labels=['Under 16','16_to_35','35_to_70','70_to_100'])))
print (ad)

       age         workclass  fnlwgt   education  education-num  \
0       39         State-gov   77516   Bachelors             13   
1       50  Self-emp-not-inc   83311   Bachelors             13   
2       38           Private  215646     HS-grad              9   
3       53           Private  234721        11th              7   
4       28           Private  338409   Bachelors             13   
...    ...               ...     ...         ...            ...   
32556   27           Private  257302  Assoc-acdm             12   
32557   40           Private  154374     HS-grad              9   
32558   58           Private  151910     HS-grad              9   
32559   22           Private  201490     HS-grad              9   
32560   52      Self-emp-inc  287927     HS-grad              9   

           marital-status         occupation   relationship   race     sex  \
0           Never-married       Adm-clerical  Not-in-family  White    Male   
1      Married-civ-spouse    Exec-manag

In [35]:
import numpy as np
co = [(adults['age'] >= 16) & (adults['age'] < 35), (adults['age'] >= 35) & (adults['age'] < 70), adults['age'] >= 70]
ch = ['young', 'adult', 'retire']
adults['AgeGroup'] = np.select(co, ch)
adults.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary AgeGroup  
0          2174             0              40  United-States  <=50K    adult  
1             0             0              13  United-States  <=50K    adult  
2             0             0              40  United-States  <=50K    adult  
3             0             0              40  United-States  <=50K    adult  
4             0             0              40           Cuba  <=50K    young

# 12-13. Определите количество зарабатывающих >50K в каждой из возрастных групп (колонка AgeGroup), а также выведите название возрастной группы, в которой чаще зарабатывают больше 50К (>50K)

In [36]:

salary_age_group = adults['AgeGroup'][adults['salary']=='>50K'].groupby(adults['AgeGroup']).count()
max_salary_age_group = adults['AgeGroup'][adults['salary']=='>50K'].groupby(adults['AgeGroup']).count().sort_values(ascending=False).head(1)

print(salary_age_group)
print(max_salary_age_group)

AgeGroup
adult     6240
retire     113
young     1488
Name: AgeGroup, dtype: int64
AgeGroup
adult    6240
Name: AgeGroup, dtype: int64


# 14. Сгруппируйте людей по типу занятости (колонка occupation) и определите количество людей в каждой группе. После чего напишите функциюю фильтрации filter_func, которая будет возвращать только те группы, в которых средний возраст (колонка age) не больше 40 и в которых все работники отрабатывают более 5 часов в неделю (колонка hours-per-week)

In [37]:
adults['occupation'] = adults['occupation'].replace('?', 'Unknown')
ocupation_data = adults['occupation'].value_counts()
print(ocupation_data)

Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Unknown              1843
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: occupation, dtype: int64
